### PR TITLE
use a reentrant lock as the mutex for initializing/reading the global instance

### LINKED ISSUE #1060: OpenTelemetry singleton shouldn't synchronize on `.class`
> you should never synchronize on something that is publicly accessible like a class object. If we need to synchronized, please create an internal lock Object to synchronize on.

_Originally posted by @jkwatson in https://github.com/open-telemetry/opentelemetry-java/pull/1058#discussion_r401045588_

### PR DESCRIPTION
**Pull request by jkwatson**
resolves #1060

### CODE REVIEW

**Comment by Oberon00** [comment]
Why not a simple `private static final Object lock = new Object()` to be used with `synchronized`? Wouldn't that be faster? Should we add a jmh benchmark?

File: api/src/main/java/io/opentelemetry/OpenTelemetry.java [hunk 1]
```diff
@@ -128,14 +133,16 @@ public static void setPropagators(ContextPropagators propagators) {
 
   /** Lazy loads an instance. */
   private static OpenTelemetry getInstance() {
-    if (instance == null) {
-      synchronized (OpenTelemetry.class) {
-        if (instance == null) {
-          instance = new OpenTelemetry();
-        }
+    mutex.lock();
```
Comment (carlosalberto) [inline api/src/main/java/io/opentelemetry/OpenTelemetry.java:136]: Any reason to remove the 'pre-check' without the lock?

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/OpenTelemetry.java:136]: unfortunately, errorprone will complain if you try to do this. I guess I could remove the @GuardedBy on the field?

**Comment by jkwatson** [comment]
> Why not a simple `private static final Object lock = new Object()` to be used with `synchronized`? Wouldn't that be faster? Should we add a jmh benchmark?

It might or might not be faster. One nice thing about using the `ReentrantLock` is that it's actually synchronized-free under the hood. I think it's implemented with some sort of queuing mechanism internally. I can add a quick benchmark to this PR, though.

File: api/src/main/java/io/opentelemetry/OpenTelemetry.java (hunk 1 above)
Comment (Oberon00) [reply api/src/main/java/io/opentelemetry/OpenTelemetry.java:136]: Another argument for the plain `Object` I guess.

Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/OpenTelemetry.java:136]: errorprone will do the same complaining in that case as well.

Comment (Oberon00) [reply api/src/main/java/io/opentelemetry/OpenTelemetry.java:136]: Ah, I see it does not understand double-checked locking!

**Comment by Oberon00** [comment]
`synchronized` only supports a subset of usages of `ReentrantLock` so it *could* be implemented faster, at least if the JIT/compiler can determine that we never use wait/notify on it.

**Comment by jkwatson** [comment]
> `synchronized` only supports a subset of usages of `ReentrantLock` so it _could_ be implemented faster, at least if the JIT/compiler can determine that we never use wait/notify on it.

My guess is that it'll probably not matter in the real world. So, we should probably go with whatever is easiest to read, and if someone can come up with a real-world use-case where the difference matters, then we can address it then. I do know that direct use of `synchronized` should be discouraged at this point, in favor of the higher level abstractions in java.util.concurrent, which is why I chose this particular implementation.

**Comment by jkwatson** [comment]
@carlosalberto : I switched back to synchronizing on a private instance for now.

**Comment by carlosalberto** [comment]
@jkwatson Thanks. Lets go with it for now. We can always revisit using a mutex instead if needed.

**Pull request merged**